backend/app/hotdog/main.py

```diff
@@ -0,0 +1,249 @@
+    ):
+        return Bundle(
+            head=head if head is not None else self.head,
+            status=status if status is not None else self.status,
+        )
+
+    @staticmethod
+    def default():
+        return Bundle(
+            head=Head(
+                corgi_ref="main",
+                enki_ref="main",
+                corgi_modified=time(),
+                enki_modified=time(),
+            ),
+            status=Status(enki_status=StatusType.running),
+        )
+
+
+class BundleManager:
+    __slots__ = ("_bundle", "_bundle_path")
+
+    def __init__(self, bundle_path: str):
+        self._bundle: Optional[Bundle] = None
+        self._bundle_path = bundle_path
+
+    def _save(self, bundle: Bundle):
+        with open(self._bundle_path, "w") as fout:
+            fout.write(bundle.json())
+        self._bundle = bundle
+
+    def save(
+        self, head: Optional[Head] = None, status: Optional[Status] = None
+    ):
+        self._save(self.bundle.update(head, status))
+
+    @property
+    def bundle(self) -> Bundle:
+        if self._bundle is not None:
+            return self._bundle
+
+        bundle = Bundle.default()
+        try:
+            default_keys = Bundle.__fields__.keys()
+            with open(self._bundle_path) as fin:
+                saved = json.load(fin)
+            bundle = bundle.update(
+                **{k: v for k, v in saved.items() if k in default_keys}
+            )
+        except Exception:
+            self._save(bundle)
+
+        return bundle
+
+
+# NOTE: These global variables will not work if multiple instances are running
+CORGI_REPO = "https://github.com/openstax/corgi"
+FRONTEND_PORT = os.environ.get('HOTDOG_FRONTEND_PORT', '8080')
+FRONTEND_HOTDOG = f"http://frontend:{FRONTEND_PORT}/checkout"
+REPO_PATH = "/corgi"
+BACKEND_REPO_DIR = os.path.join(REPO_PATH, "backend", "app")
+BACKEND_DIR = "/app"
+REF_REGEX = re.compile(r"^[a-zA-Z0-9_./-]+$")
+BUNDLE_MGR = BundleManager("/persist/bundle.json")
+
+
+def sh(cmd: str, ignore_exitcode=False):
+    p = run(split(cmd), stdout=PIPE, stderr=PIPE)
+    if not ignore_exitcode and p.returncode != 0:
+        raise Exception(p.stderr.decode())
+    return (p, p.stdout.decode(), p.stderr.decode())
+
+
+def scoped_git(cmd, scope=REPO_PATH, ignore_exitcode=False):
+    return sh(f"git -C {scope} {cmd}", ignore_exitcode)
+
+
+def is_pullable(ref):
+    return any(
+        line.endswith(ref)
+        for line in scoped_git("show-ref")[1].strip().splitlines()
+        if "refs/heads/" in line or "refs/remotes" in line
+    )
+
+
+def _corgi_checkout(ref):
+    scoped_git("fetch")
+    scoped_git(f"checkout {ref}")
+    if is_pullable(ref):
+        scoped_git("pull --rebase")
+    copytree(
+        BACKEND_REPO_DIR,
+        BACKEND_DIR,
+        dirs_exist_ok=True,
+        ignore=ignore_patterns("hotdog*"),
+    )
+    response = httpx.post(FRONTEND_HOTDOG, json={"ref": ref})
```

Comment:
What does this accomplish? Maybe triggers the FE to reload? (maybe just a note for myself to follow where this leads)